File: crates/orbit-web/assets/dashboard/distributed.js
// Orbit dashboard distributed-drain provenance and owner handoff actions
// (ORB-12516). Pure vanilla JS ES module, no build step.
//
// This module renders claim provenance into the *existing* task and run views
// and offers the owner's approve / revoke / recover actions there. It is not a
// distributed console: there is no tab, no route and no navigation entry, so the
// incomplete feature gains no public surface of its own — the panel appears only
// where the owner actually holds a claim for the task being looked at.
//
// Three rules shape every string below.
//
//  1. Unknown stays unknown. A run or artifact recorded before execution
//     provenance existed reads as "unknown", never as "the owner".
//  2. A diagnostic is not a verdict. An elapsed reservation, a claim's age and
//     an absent owner-local run are things to look at, not proof the attempt
//     died and not revocation. V1 has manual reclamation for exactly that
//     reason.
//  3. Say what actually happened. `review` means a delivery handoff is waiting
//     for completion authority, not that a reviewer approved anything; merged
//     means the candidate is on the landing branch, not that it is deployed.
//
// The buttons are a convenience, never the boundary: every action carries the
// exact identity the operator was shown, and the owner re-decides authority,
// currentness and merge certainty inside the transaction that would change
// anything.

import { el, fetchJson, postJson, makeToggleRow, isAggregateView } from './common.js';

const CONSOLE_PATH = "/api/distributed/claims";

// One read per view, shared by however many rows are expanded. Invalidated on
// every action so a decision is never rendered from the state that preceded it.
let cachedConsole = null;
let inflightConsole = null;

/// Drop the memoized read. Called after any owner action and by the workspace
/// selector, since claim state is per-workspace.
export function invalidateDistributedConsole() {
  cachedConsole = null;
  inflightConsole = null;
}

export function peekDistributedConsole() {
  return cachedConsole;
}

export function loadDistributedConsole({ force = false } = {}) {
  // Claim state is per-workspace. The aggregate view has no concrete workspace
  // to scope to — the endpoint would 400 — so answer as "nothing here" rather
  // than painting a failure across every task detail.
  if (isAggregateView()) return Promise.resolve(null);
  if (force) invalidateDistributedConsole();
  if (cachedConsole) return Promise.resolve(cachedConsole);
  if (!inflightConsole) {
    inflightConsole = fetchJson(CONSOLE_PATH)
      .then((payload) => {
        cachedConsole = payload || {};
        inflightConsole = null;
        return cachedConsole;
      })
      .catch((error) => {
        inflightConsole = null;
        throw error;
      });
  }
  return inflightConsole;
}

export function claimsForTask(payload, taskId) {
  const claims = payload && Array.isArray(payload.claims) ? payload.claims : [];
  return claims.filter((claim) => claim && claim.task_id === taskId);
}

// --- shared renderers -------------------------------------------------------

/// Host-qualified execution provenance, or an explicit unknown.
///
/// `machine_id` is the stable identity; `host_id` rides along for display and
/// may be renamed, so it is never shown alone.
export function formatExecutionLocation(location) {
  if (!location || location.known !== true || !location.machine_id) {
    return "unknown — recorded before execution provenance was tracked";
  }
  return location.host_id
    ? `machine ${location.machine_id} · host ${location.host_id}`
    : `machine ${location.machine_id}`;
}

/// The execution-provenance cell shared by the run meta grid and the task
/// detail's run line.
export function buildExecutionProvenance(location) {
  const known = !!(location && location.known === true && location.machine_id);
  return el("span", {
    class: known ? "exec-origin" : "exec-origin unknown",
    text: formatExecutionLocation(location),
    title: known
      ? "Where this ran. Nothing is inferred from hostname, cwd or SSH target."
      : "No execution machine was recorded for this row. Unknown is not the owner.",
  });
}

function line(label, value, opts = {}) {
  const row = el("div", { class: opts.class ? `claim-line ${opts.class}` : "claim-line" }, [
    el("span", { class: "label", text: label }),
    typeof value === "string" || value == null
      ? el("span", { class: "value", text: value == null ? "—" : value })
      : value,
  ]);
  if (opts.note) row.appendChild(el("div", { class: "claim-note", text: opts.note }));
  return row;
}

function shortCommit(value) {
  return typeof value === "string" && value.length > 12 ? value.slice(0, 12) : (value || "—");
}

// --- claim panel ------------------------------------------------------------

/// Render one claim: provenance, phase, footprint, reservation, handoff and the
/// owner actions the caller wires.
///
/// Pure with respect to the DOM it creates — it reads only its arguments — so
/// the behavior scenarios can drive it without a server.
export function buildClaimPanel(claim, capabilities, handlers = {}) {
  const panel = el("div", { class: "claim-panel" });
  panel.setAttribute("data-claim-id", claim.claim_id || "");
  panel.setAttribute("data-phase", claim.phase || "");

  panel.appendChild(line("execution", buildExecutionProvenance(claim.executed_on)));

  const run = claim.bound_run;
  if (run && run.run_id) {
    if (claim.bound_run_navigable) {
      const link = el("a", { class: "value", text: run.run_id });
      link.href = `#runs?run_id=${encodeURIComponent(run.run_id)}`;
      panel.appendChild(line("bound run", link));
    } else {
      // No owner-local run exists for a run that executed elsewhere. Naming the
      // host is the honest answer; an owner-local link would resolve against
      // this checkout's job store and open the wrong thing or nothing.
      panel.appendChild(
        line("bound run", `${run.run_id} on machine ${run.machine_id}`, {
          class: "remote",
          note: claim.inspect_on || "inspect this run on its execution host",
        }),
      );
    }
  } else {
    panel.appendChild(
      line("bound run", "none yet", {
        note: "a claim exists before its leaf run does; an absent run is not proof the attempt died",
      }),
    );
  }

  panel.appendChild(line("phase", `${claim.phase} — ${claim.phase_summary || ""}`.trim()));
  panel.appendChild(line("last event", claim.last_event));

  const reservation = claim.reservation || {};
  panel.appendChild(
    line(
      "reservation",
      reservation.expired ? `expired ${reservation.expires_at}` : `expires ${reservation.expires_at}`,
      { class: reservation.expired ? "expired" : "", note: reservation.note },
    ),
  );

  const footprint = Array.isArray(claim.footprint) ? claim.footprint : [];
  const files = el("div", { class: "claim-footprint" });
  for (const entry of footprint) files.appendChild(el("div", { class: "mono", text: entry }));
  panel.appendChild(
    line("footprint", files, {
      note: claim.footprint_protected
        ? `${footprint.length} frozen selector(s), still protected`
        : `${footprint.length} selector(s), no longer protecting files`,
    }),
  );

  if (claim.landing_invalidated) {
    panel.appendChild(
      el("div", { class: "claim-warning", text: "landing authority for this claim was invalidated" }),
    );
  }
  if (claim.unresolved_merge_intent) {
    panel.appendChild(uncertainMergeBanner(claim.unresolved_merge_intent));
  }

  if (claim.handoff) panel.appendChild(buildHandoffPanel(claim.handoff));

  const actions = buildClaimActions(claim, capabilities, handlers);
  if (actions) panel.appendChild(actions);
  return panel;
}

/// An external merge whose reply was lost. Until it is reconciled against the
/// provider, neither revocation nor recovery may proceed: the row in the
/// database cannot cancel a request the provider may already have applied.
function uncertainMergeBanner(intentId) {
  return el("div", { class: "claim-warning uncertain-merge" }, [
    el("span", { class: "label", text: "uncertain merge intent" }),
    el("span", { class: "value mono", text: intentId }),
    el("div", {
      class: "claim-note",
      text: "a merge was recorded as sent and its outcome is unknown; reconcile it against the provider before revoking or recovering — revocation and recovery are refused until then",
    }),
  ]);
}

function buildHandoffPanel(handoff) {
  const wrap = el("div", { class: "handoff-panel" });
  wrap.setAttribute("data-handoff-id", handoff.handoff_id || "");
  wrap.appendChild(el("h5", { text: "delivery handoff" }));

  const candidate = handoff.candidate || {};
  const delivery = candidate.delivery || {};
  wrap.appendChild(line("repository", candidate.repository));
  wrap.appendChild(
    line("branches", `${candidate.source_branch} → ${candidate.base_branch} (lands on ${candidate.landing_branch})`),
  );
  wrap.appendChild(
    line(
      "candidate",
      `${shortCommit(candidate.candidate && candidate.candidate.commit)} on base ${shortCommit(candidate.base && candidate.base.commit)}`,
    ),
  );
  wrap.appendChild(line("delivery", deliveryLabel(delivery)));

  const review = handoff.review || {};
  // The typed not-required disposition, said plainly. `review` as a task status
  // means delivery is awaiting completion authority; no reviewer ran.
  wrap.appendChild(
    line("review", `${review.disposition || "?"} (policy ${review.policy || "?"})`, {
      class: "review-disposition",
      note: review.summary || "review not required — this is not a code review",
    }),
  );

  const validation = Array.isArray(handoff.validation) ? handoff.validation : [];
  const required = Array.isArray(handoff.required_commands) ? handoff.required_commands : [];
  const logs = el("div", { class: "handoff-validation" });
  for (const entry of validation) {
    logs.appendChild(el("div", { class: "mono", text: `${entry.path} · ${shortCommit(entry.sha256)}` }));
  }
  wrap.appendChild(
    line("validation", logs, {
      note: `${validation.length} captured log(s) for ${required.length} owner-required command(s): ${required.join(", ") || "none"}`,
    }),
  );

  const authority = handoff.authority || {};
  wrap.appendChild(
    line("completion authority", authority.state, {
      class: `authority-${authority.state}`,
      note: authority.summary,
    }),
  );

  const landing = handoff.landing || {};
  wrap.appendChild(
    line("landing", landing.state, {
      class: `landing-${landing.state}`,
      note: landing.summary,
    }),
  );
  if (landing.evidence) wrap.appendChild(line("landing evidence", landing.evidence));
  return wrap;
}

function deliveryLabel(delivery) {
  switch (delivery && delivery.kind) {
    case "pull_request":
      return `pull request #${delivery.number}`;
    case "local_candidate":
      return "local candidate (owner performs the merge)";
    case "already_landed":
      return `already landed in ${shortCommit(delivery.covering_commit)}`;
    default:
      return "—";
  }
}

// --- owner actions ----------------------------------------------------------

function capabilityFor(capabilities, key) {
  const entry = capabilities && capabilities[key];
  return entry && typeof entry === "object" ? entry : { authorized: false, reason: "unavailable" };
}

function buildClaimActions(claim, capabilities, handlers) {
  const handoff = claim.handoff;
  const authority = (handoff && handoff.authority) || {};
  const row = el("div", { class: "claim-actions" });
  let any = false;

  const add = (key, label, capabilityKey, onRun, opts = {}) => {
    const capability = capabilityFor(capabilities, capabilityKey);
    any = true;
    if (!capability.authorized) {
      // Explain rather than hide: an operator who cannot act should learn why,
      // and the backend refuses the call regardless of what is rendered here.
      row.appendChild(
        el("span", {
          class: "claim-action-denied",
          text: `${label} unavailable`,
          title: capability.reason || "not authorized",
        }),
      );
      return;
    }
    const button = el("button", { class: `claim-action ${key}`, text: label });
    button.type = "button";
    button.setAttribute("data-action", key);
    if (opts.reasonRequired) {
      const input = el("input", { class: "claim-reason" });
      input.type = "text";
      input.placeholder = opts.reasonPlaceholder || "reason";
      input.setAttribute("aria-label", `${label} reason`);
      input.setAttribute("data-reason-for", key);
      row.appendChild(input);
      button.addEventListener("click", () => onRun(input.value));
    } else {
      button.addEventListener("click", () => onRun(null));
    }
    row.appendChild(button);
  };

  if (handoff && claim.phase === "handed_off" && authority.state === "not_authorized") {
    add("approve", "Approve handoff", "handoff_approve", () => handlers.onApprove && handlers.onApprove(claim));
  }
  if (handoff && authority.state === "authorized") {
    add(
      "revoke",
      "Revoke authority",
      "handoff_revoke",
      (reason) => handlers.onRevoke && handlers.onRevoke(claim, reason),
      { reasonRequired: true, reasonPlaceholder: "why authority is withdrawn" },
    );
  }
  if (claim.unsettled) {
    add(
      "recover",
      "Recover claim",
      "claim_recover",
      (reason) => handlers.onRecover && handlers.onRecover(claim, reason),
      { reasonRequired: true, reasonPlaceholder: "why this attempt is over" },
    );
  }

  if (!any) return null;
  row.appendChild(
    el("div", {
      class: "claim-note",
      text: "Approving records completion authority for this exact candidate; it does not merge. Recovery fences the attempt and is never automatic — there is no heartbeat.",
    }),
  );
  return row;
}

/// A replay identity for one decision. A retried POST with the same value
/// replays the stored outcome instead of recording a second one.
function newRequestId() {
  const random = Math.random().toString(36).slice(2, 10);
  return `dash-${Date.now().toString(36)}-${random}`;
}

function expectedCandidate(claim) {
  const candidate = (claim.handoff && claim.handoff.candidate) || {};
  return {
    expected_candidate_commit: (candidate.candidate && candidate.candidate.commit) || "",
    expected_base_commit: (candidate.base && candidate.base.commit) || "",
  };
}

export function approveHandoff(claim) {
  return postJson(`/api/distributed/handoffs/${encodeURIComponent(claim.handoff.handoff_id)}/approve`, {
    ...expectedCandidate(claim),
    request_id: newRequestId(),
  });
}

export function revokeHandoff(claim, reason) {
  return postJson(`/api/distributed/handoffs/${encodeURIComponent(claim.handoff.handoff_id)}/revoke`, {
    ...expectedCandidate(claim),
    reason: reason || "",
    request_id: newRequestId(),
  });
}

export function recoverClaim(claim, reason, status = "blocked") {
  return postJson(`/api/distributed/claims/${encodeURIComponent(claim.claim_id)}/recover`, {
    expected_phase: claim.phase,
    status,
    reason: reason || "",
    request_id: newRequestId(),
  });
}

// --- mounting into the task detail -----------------------------------------

/// Fill `container` with the distributed panel for `taskId`, reading the console
/// once and re-reading it after every action.
///
/// The container is left empty when this workspace holds no claim for the task,
/// which is the normal case: the block simply does not appear.
export function mountTaskClaimPanel(container, taskId, { onTaskChanged, onContent } = {}) {
  const announce = (rendered) => {
    if (onContent) onContent(rendered);
    return rendered;
  };
  const render = (payload) => {
    container.textContent = "";
    if (payload && payload.owner_workspace === false) {
      // A replica checkout: claim state lives on the owner machine. Say so
      // rather than rendering an empty panel that reads as "no claims".
      container.appendChild(
        el("div", {
          class: "claim-note",
          text: payload.refusal_detail || "this checkout is a replica; the owner machine holds claim state",
        }),
      );
      return announce(true);
    }
    const claims = claimsForTask(payload, taskId);
    // The common case: this workspace holds no claim for this task, so the
    // block does not appear at all.
    if (claims.length === 0) return announce(false);
    const capabilities = (payload && payload.capabilities) || {};
    for (const claim of claims) {
      container.appendChild(
        buildClaimPanel(claim, capabilities, {
          onApprove: (target) => act(() => approveHandoff(target)),
          onRevoke: (target, reason) => act(() => revokeHandoff(target, reason)),
          onRecover: (target, reason) => act(() => recoverClaim(target, reason)),
        }),
      );
    }
    return announce(true);
  };

  const feedback = (text, kind) => {
    const note = el("div", { class: `claim-feedback ${kind}`, text });
    note.setAttribute("role", "status");
    container.insertBefore(note, container.firstChild);
  };

  const act = (run) =>
    run()
      .then((body) => {
        if (onTaskChanged) onTaskChanged();
        // Re-read before reporting: the operator should see the decision and
        // the state it produced together, never the state that preceded it.
        return refresh().then(() => feedback(actionSummary(body), "ok"));
      })
      .catch((error) => {
        // A refused action is the point of the backend check, not a bug in it.
        // Stale state re-reads before the operator decides again; an uncertain
        // merge is left alone until it is reconciled.
        const stale = error && (error.code === "stale_claim" || error.code === "handoff_not_current");
        const message = error && error.message ? error.message : String(error);
        if (!stale) {
          feedback(message, error && error.code === "uncertain_merge_intent" ? "uncertain" : "error");
          return;
        }
        return refresh().then(() => feedback(`${message} — refreshed`, "stale"));
      });

  const failed = (error) => {
    container.textContent = "";
    container.appendChild(
      el("div", { class: "claim-feedback error", text: `claim state unavailable: ${error.message || error}` }),
    );
    return announce(true);
  };

  const refresh = () => loadDistributedConsole({ force: true }).then(render).catch(failed);

  const cached = peekDistributedConsole();
  if (cached) return Promise.resolve(render(cached));
  container.appendChild(el("div", { class: "claim-note", text: "reading claim state…" }));
  return loadDistributedConsole().then(render).catch(failed);
}

function actionSummary(body) {
  const result = (body && body.result) || {};
  if (result.handoff_id) {
    return `handoff ${result.handoff_id}: claim ${result.phase}, task ${result.task_status}`;
  }
  return `claim ${result.claim_id || "?"}: ${result.phase || "?"}, task ${result.task_status || "?"}`;
}

/// Collapsible wrapper matching the task detail's other field blocks, so the
/// panel is reachable by keyboard the same way every other block is.
export function buildDistributedBlock(taskId, opts = {}) {
  const block = el("div", { class: "field-block collapsible distributed-block" });
  // Hidden until the read says this task actually has a claim, so an ordinary
  // single-host task detail is unchanged.
  block.style.display = "none";
  const heading = el("h4", {}, [el("span", { class: "field-title", text: "distributed execution" })]);
  const body = el("div", { class: "claim-body" });
  makeToggleRow(heading, {
    expanded: true,
    onToggle: (event) => {
      event.stopPropagation();
      const collapsed = block.classList.toggle("collapsed");
      heading.setAttribute("aria-expanded", String(!collapsed));
    },
  });
  block.appendChild(heading);
  block.appendChild(body);
  mountTaskClaimPanel(body, taskId, {
    ...opts,
    onContent: (rendered) => {
      block.style.display = rendered ? "" : "none";
      if (opts.onContent) opts.onContent(rendered);
    },
  });
  return block;
}

File: crates/orbit-web/src/api/distributed.rs
//! Distributed-drain claim provenance and the owner's handoff actions
//! [ORB-12516].
//!
//! # What this adapter is
//!
//! An HTTP projection of owner-domain lifecycle APIs that already exist. It
//! reuses the dashboard's task and run views rather than adding a console of
//! its own, and it holds no lifecycle state: every read is
//! [`OrbitRuntime::distributed_claim_console`] and every write ends in the same
//! canonical approve / revoke / recover transaction the owner landing consumer
//! uses.
//!
//! # Where authority comes from
//!
//! The same place every other dashboard action's does: a
//! [`GovernedOperation`](orbit_common::governance::authorization::GovernedOperation)
//! row resolved through the shared chokepoint. There is no second authority
//! model here, and no action is gated by a disabled button — the capability
//! projection in the read response exists so the UI can *explain* a refusal it
//! would get anyway, and each mutation re-resolves it server-side.
//!
//! Three more fences sit behind that one, and none of them is optional:
//!
//! - **Replica.** Coordination writes are refused in a replica checkout by
//!   Core's `ensure_coordination_task_write_permitted`. The read answers such a
//!   checkout honestly (owner machine named, no claims) instead of erroring.
//! - **Stale.** Every mutation carries the identity the operator was looking at
//!   — the candidate/base commit pair for a handoff, the phase for a claim. A
//!   mismatch is a 409 the client resolves by refreshing, decided against the
//!   owner's stored record rather than against the browser's copy of it.
//! - **Uncertain merge.** A merge intent whose reply was lost blocks revocation
//!   and recovery until it is reconciled against the provider. A database row
//!   cannot cancel a request GitHub may already have applied.
//!
//! # What it deliberately does not do
//!
//! No host registration, demotion, or fleet listing; no automatic reclamation;
//! no landing trigger. Recovery is here because there is no heartbeat, so the
//! decision that an attempt is over belongs to a human.

use std::time::Instant;

use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Json, Response};
use orbit_common::governance::authorization::{
    DASHBOARD_CLAIM_RECOVER, DASHBOARD_HANDOFF_APPROVE, DASHBOARD_HANDOFF_REVOKE,
};
use orbit_core::application::review::{ExpectedCandidate, HandoffConsoleRefusal};
use orbit_core::{OrbitError, OrbitRuntime, TaskStatus};
use serde::Deserialize;
use serde_json::{Value, json};

use super::routines::{
    OperationsQuery, action_capability, authorization_denied, authorized_caller,
    explicit_workspace, record_operation_audit,
};
use super::{bad_request, blocking, server_error, validate_id};
use crate::state::{DashboardState, Ws};

/// Actor recorded for a dashboard-authored owner decision. The dashboard is a
/// human surface, so approval is attributed to a human even when the server
/// process itself runs inside a managed run whose ambient identity is a model.
const OPERATOR_ACTOR: &str = "human";

/// Longest accepted replay identity. Long enough for a UUID or a digest,
/// short enough that a client cannot use the mutation key as storage.
const MAX_REQUEST_ID: usize = 128;
/// Longest accepted operator reason, matching what the store records as a
/// status note.
const MAX_REASON: usize = 2000;

#[derive(Debug, Deserialize)]
pub(super) struct ApproveHandoffRequest {
    /// The candidate the operator was shown. Compared against the owner's
    /// accepted handoff and then discarded.
    expected_candidate_commit: String,
    expected_base_commit: String,
    /// Replay identity: retrying the same decision returns the stored result
    /// instead of recording a second one.
    request_id: String,
}

#[derive(Debug, Deserialize)]
pub(super) struct RevokeHandoffRequest {
    expected_candidate_commit: String,
    expected_base_commit: String,
    reason: String,
    request_id: String,
}

#[derive(Debug, Deserialize)]
pub(super) struct RecoverClaimRequest {
    /// The phase the operator was shown. A claim that moved on is refused.
    expected_phase: String,
    /// `blocked` for diagnosis or `backlog` for retry — the operator's choice,
    /// never inferred.
    status: String,
    reason: String,
    request_id: String,
}

/// `GET /api/distributed/claims` — read-only owner inspection.
///
/// Creates nothing: no receipt, no reservation, no claim, no task transition.
/// A replica checkout is answered rather than refused, so switching the
/// workspace selector renders an honest empty view instead of a fault.
pub(super) async fn list_claims(State(state): State<DashboardState>, Ws(runtime): Ws) -> Response {
    let operator_session = state.operator_session();
    match blocking("distributed claim console", move || {
        runtime.distributed_claim_console()
    })
    .await
    {
        Ok(mut console) => {
            if let Some(object) = console.as_object_mut() {
                object.insert("capabilities".to_string(), capabilities(operator_session));
            }
            Json(console).into_response()
        }
        Err(response) => *response,
    }
}

/// `POST /api/distributed/handoffs/:id/approve` — record completion authority
/// for one exact candidate.
///
/// Approving does not merge. It records an immutable, candidate-scoped
/// authorization and the durable landing-start request the owner landing job
/// consumes; the merge itself re-checks the authorization, the candidate and
/// the digest-pinned validation evidence in its own transaction.
pub(super) async fn approve_handoff_action(
    State(state): State<DashboardState>,
    Query(query): Query<OperationsQuery>,
    Ws(runtime): Ws,
    Path(handoff_id): Path<String>,
    Json(body): Json<ApproveHandoffRequest>,
) -> Response {
    let handoff_id = match validate_id(&handoff_id) {
        Ok(id) => id.to_string(),
        Err(message) => return bad_request(message),
    };
    let request_id = match validate_request_id(&body.request_id) {
        Ok(request_id) => request_id,
        Err(message) => return bad_request(message),
    };
    let expected = ExpectedCandidate {
        candidate_commit: body.expected_candidate_commit,
        base_commit: body.expected_base_commit,
    };
    let arguments = json!({
        "handoff_id": handoff_id,
        "candidate": expected.candidate_commit,
        "base": expected.base_commit,
    });
    let audited = handoff_id.clone();
    run_owner_action(
        state,
        query,
        runtime,
        &DASHBOARD_HANDOFF_APPROVE,
        "handoff.approve",
        audited,
        arguments,
        move |runtime| {
            runtime.approve_handoff_as_operator(
                &handoff_id,
                &expected,
                OPERATOR_ACTOR,
                &format!("dashboard-approve:{handoff_id}:{request_id}"),
            )
        },
    )
    .await
}

/// `POST /api/distributed/handoffs/:id/revoke` — withdraw completion authority.
///
/// The task stays in `review`: revocation cancels the pending landing request
/// and records its own immutable audit row. It does not decide what happens to
/// the work, and it is refused while an external merge intent is unresolved.
pub(super) async fn revoke_handoff_action(
    State(state): State<DashboardState>,
    Query(query): Query<OperationsQuery>,
    Ws(runtime): Ws,
    Path(handoff_id): Path<String>,
    Json(body): Json<RevokeHandoffRequest>,
) -> Response {
    let handoff_id = match validate_id(&handoff_id) {
        Ok(id) => id.to_string(),
        Err(message) => return bad_request(message),
    };
    let request_id = match validate_request_id(&body.request_id) {
        Ok(request_id) => request_id,
        Err(message) => return bad_request(message),
    };
    let reason = match validate_reason(&body.reason, "revoking completion authority") {
        Ok(reason) => reason,
        Err(message) => return bad_request(message),
    };
    let expected = ExpectedCandidate {
        candidate_commit: body.expected_candidate_commit,
        base_commit: body.expected_base_commit,
    };
    let arguments = json!({
        "handoff_id": handoff_id,
        "candidate": expected.candidate_commit,
        "base": expected.base_commit,
        "reason": reason,
    });
    let audited = handoff_id.clone();
    run_owner_action(
        state,
        query,
        runtime,
        &DASHBOARD_HANDOFF_REVOKE,
        "handoff.revoke",
        audited,
        arguments,
        move |runtime| {
            runtime.revoke_handoff_as_operator(
                &handoff_id,
                &expected,
                OPERATOR_ACTOR,
                &reason,
                &format!("dashboard-revoke:{handoff_id}:{request_id}"),
            )
        },
    )
    .await
}

/// `POST /api/distributed/claims/:id/recover` — deliberate recovery.
///
/// Fences the attempt, invalidates any pending landing authority, releases the
/// reservation and moves the task, all in one transaction. Nothing here infers
/// that an attempt died: age, an elapsed reservation and an absent owner-local
/// run are diagnostics the read surfaces, and this is the operator acting on
/// them.
pub(super) async fn recover_claim_action(
    State(state): State<DashboardState>,
    Query(query): Query<OperationsQuery>,
    Ws(runtime): Ws,
    Path(claim_id): Path<String>,
    Json(body): Json<RecoverClaimRequest>,
) -> Response {
    let claim_id = match validate_id(&claim_id) {
        Ok(id) => id.to_string(),
        Err(message) => return bad_request(message),
    };
    let request_id = match validate_request_id(&body.request_id) {
        Ok(request_id) => request_id,
        Err(message) => return bad_request(message),
    };
    let reason = match validate_reason(&body.reason, "claim recovery") {
        Ok(reason) => reason,
        Err(message) => return bad_request(message),
    };
    let status = match body.status.as_str() {
        "blocked" => TaskStatus::Blocked,
        "backlog" => TaskStatus::Backlog,
        other => {
            return bad_request(format!(
                "recovery moves a task to 'blocked' for diagnosis or 'backlog' for retry, not \
                 '{other}'"
            ));
        }
    };
    let expected_phase = body.expected_phase;
    let arguments = json!({
        "claim_id": claim_id,
        "expected_phase": expected_phase,
        "status": body.status,
        "reason": reason,
    });
    let audited = claim_id.clone();
    run_owner_action(
        state,
        query,
        runtime,
        &DASHBOARD_CLAIM_RECOVER,
        "claim.recover",
        audited,
        arguments,
        move |runtime| {
            runtime.recover_claim_as_operator(
                &claim_id,
                &expected_phase,
                status,
                OPERATOR_ACTOR,
                &reason,
                &format!("dashboard-recover:{claim_id}:{request_id}"),
            )
        },
    )
    .await
}

/// The shared body of all three mutations: authorize, require an explicit
/// workspace, run the domain transaction off the async pool, audit the outcome,
/// and translate a refusal into the code the client acts on.
#[allow(clippy::too_many_arguments)]
async fn run_owner_action<F>(
    state: DashboardState,
    query: OperationsQuery,
    runtime: std::sync::Arc<OrbitRuntime>,
    operation: &'static orbit_common::governance::authorization::GovernedOperation,
    audit_name: &'static str,
    target: String,
    arguments: Value,
    action: F,
) -> Response
where
    F: FnOnce(&OrbitRuntime) -> Result<Value, OrbitError> + Send + 'static,
{
    // Workspace selection first: an action aimed at the wrong workspace is a
    // malformed request, not a denial, and auditing it as one would be a lie.
    let workspace = match explicit_workspace(&query) {
        Ok(workspace) => workspace.to_string(),
        Err(rejection) => return rejection.into_response(),
    };
    let started = Instant::now();
    let caller = match authorized_caller(operation, state.operator_session()) {
        Ok(caller) => caller,
        Err(denial) => {
            record_operation_audit(
                &runtime,
                &workspace,
                audit_name,
                &target,
                "",
                &arguments,
                None,
                Some(&denial),
                None,
                started,
            );
            return authorization_denied(denial);
        }
    };
    let acting = runtime.clone();
    let outcome = tokio::task::spawn_blocking(move || action(&acting)).await;
    match outcome {
        Ok(Ok(value)) => {
            record_operation_audit(
                &runtime,
                &workspace,
                audit_name,
                &target,
                "",
                &arguments,
                Some(&caller),
                None,
                None,
                started,
            );
            Json(json!({ "ok": true, "result": value })).into_response()
        }
        Ok(Err(error)) => {
            let response = refusal_response(&error);
            record_operation_audit(
                &runtime,
                &workspace,
                audit_name,
                &target,
                "",
                &arguments,
                Some(&caller),
                None,
                Some(&error.to_string()),
                started,
            );
            response
        }
        Err(join_err) => server_error(OrbitError::Execution(format!(
            "{audit_name} panicked: {join_err}"
        ))),
    }
}

/// Translate an owner-domain refusal into the status and code the dashboard
/// acts on.
///
/// The three conflict codes are distinct because the operator's next move
/// differs: refresh the view, reconcile the merge, or act on the owner machine.
/// Collapsing them into one 500 would hide exactly the state this surface
/// exists to show.
fn refusal_response(error: &OrbitError) -> Response {
    let message = error.to_string();
    match HandoffConsoleRefusal::classify(error) {
        Some(refusal @ HandoffConsoleRefusal::ReplicaCheckout) => (
            StatusCode::FORBIDDEN,
            Json(json!({"error": message, "code": refusal.code()})),
        )
            .into_response(),
        Some(refusal @ HandoffConsoleRefusal::UncertainMerge) => (
            StatusCode::CONFLICT,
            Json(json!({
                "error": message,
                "code": refusal.code(),
                "remedy": "reconcile the recorded merge intent against the provider's actual \
                           state before revoking or recovering this claim",
            })),
        )
            .into_response(),
        Some(refusal) => (
            StatusCode::CONFLICT,
            Json(json!({
                "error": message,
                "code": refusal.code(),
                "remedy": "refresh the view: the owner's claim or handoff state changed since \
                           this action was prepared",
            })),
        )
            .into_response(),
        // A handoff the owner no longer holds reads as stale too: the operator's
        // remedy is the same refresh, and a 404 would invite a client retry loop.
        None if message.contains("is current on this owner") => (
            StatusCode::CONFLICT,
            Json(json!({
                "error": message,
                "code": HandoffConsoleRefusal::NotCurrent.code(),
                "remedy": "refresh the view: this handoff or claim is no longer current",
            })),
        )
            .into_response(),
        None => match error {
            OrbitError::InvalidInput(message) => bad_request(message.clone()),
            other => server_error(OrbitError::Execution(other.to_string())),
        },
    }
}

/// The same authorization decision each mutation enforces, projected so the UI
/// can say *why* an action is unavailable instead of silently hiding it.
fn capabilities(operator_session: bool) -> Value {
    json!({
        "handoff_approve": action_capability(&DASHBOARD_HANDOFF_APPROVE, operator_session),
        "handoff_revoke": action_capability(&DASHBOARD_HANDOFF_REVOKE, operator_session),
        "claim_recover": action_capability(&DASHBOARD_CLAIM_RECOVER, operator_session),
    })
}

fn validate_request_id(raw: &str) -> Result<String, String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Err(
            "request_id identifies this decision so a retry replays it instead of recording a \
             second one; send one"
                .to_string(),
        );
    }
    if trimmed.len() > MAX_REQUEST_ID {
        return Err(format!(
            "request_id must be at most {MAX_REQUEST_ID} characters"
        ));
    }
    validate_id(trimmed).map(ToString::to_string)
}

fn validate_reason(raw: &str, action: &str) -> Result<String, String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Err(format!("{action} requires a reason"));
    }
    if trimmed.len() > MAX_REASON {
        return Err(format!("reason must be at most {MAX_REASON} characters"));
    }
    Ok(trimmed.to_string())
}
